test_proxy/client_handler.py

```diff
@@ -0,0 +1,211 @@
+        return serialized_response
+
+    @error_safe
+    async def ReadRow(self, row_key, **kwargs):
+        table_id = kwargs.pop("table_name").split("/")[-1]
+        app_profile_id = self.app_profile_id or kwargs.get("app_profile_id", None)
+        table = self.client.get_table(self.instance_id, table_id, app_profile_id)
+        kwargs["operation_timeout"] = kwargs.get("operation_timeout", self.per_operation_timeout) or 20
+        result_row = await table.read_row(row_key, **kwargs)
+        # pack results back into protobuf-parsable format
+        if result_row:
+            return result_row.to_dict()
+        else:
+            return "None"
+
+    @error_safe
+    async def MutateRow(self, request, **kwargs):
+        from google.cloud.bigtable.data.mutations import Mutation
+        table_id = request["table_name"].split("/")[-1]
+        app_profile_id = self.app_profile_id or request.get("app_profile_id", None)
+        table = self.client.get_table(self.instance_id, table_id, app_profile_id)
+        kwargs["operation_timeout"] = kwargs.get("operation_timeout", self.per_operation_timeout) or 20
+        row_key = request["row_key"]
+        mutations = [Mutation._from_dict(d) for d in request["mutations"]]
+        await table.mutate_row(row_key, mutations, **kwargs)
+        return "OK"
+
+    @error_safe
+    async def BulkMutateRows(self, request, **kwargs):
+        from google.cloud.bigtable.data.mutations import RowMutationEntry
+        table_id = request["table_name"].split("/")[-1]
+        app_profile_id = self.app_profile_id or request.get("app_profile_id", None)
+        table = self.client.get_table(self.instance_id, table_id, app_profile_id)
+        kwargs["operation_timeout"] = kwargs.get("operation_timeout", self.per_operation_timeout) or 20
+        entry_list = [RowMutationEntry._from_dict(entry) for entry in request["entries"]]
+        await table.bulk_mutate_rows(entry_list, **kwargs)
+        return "OK"
+
+    @error_safe
+    async def CheckAndMutateRow(self, request, **kwargs):
+        from google.cloud.bigtable.data.mutations import Mutation, SetCell
+        table_id = request["table_name"].split("/")[-1]
+        app_profile_id = self.app_profile_id or request.get("app_profile_id", None)
+        table = self.client.get_table(self.instance_id, table_id, app_profile_id)
+        kwargs["operation_timeout"] = kwargs.get("operation_timeout", self.per_operation_timeout) or 20
+        row_key = request["row_key"]
+        # add default values for incomplete dicts, so they can still be parsed to objects
+        true_mutations = []
+        for mut_dict in request.get("true_mutations", []):
+            try:
+                true_mutations.append(Mutation._from_dict(mut_dict))
+            except ValueError:
+                # invalid mutation type. Conformance test may be sending generic empty request
+                true_mutations.append(SetCell("", "", "", -1))
```

Comment:
It's been a while since I looked at this, but I think this is the one spot where I couldn't get a test to pass without modifications to the test code

I think the library will raise an exception if I try to create an empty SetCell mutation, but I believe the [tests expect it](https://github.com/googleapis/cloud-bigtable-clients-test/blob/3f9a44bd6038c0a7c3c801b3829fb41fb4843abc/tests/checkandmutaterow_test.go#L49). I was going to suggest we change the example mutation in the conformance tests to one with valid data

Reply:
I see. You are right.. But somehow they other clients seems to be fine with the empty set. You can keep the line as is. I will make the test change later and let you know when it's done.